File: src/cf_ip_blacklist/http.py
from __future__ import annotations

import time
from collections.abc import Callable
from typing import Any

import httpx

from .errors import CloudflareError

RETRYABLE_STATUS = {408, 429, 500, 502, 503, 504}


def request_with_retry(
    client: httpx.Client,
    method: str,
    url: str,
    max_retries: int,
    sleeper: Callable[[float], None] = time.sleep,
    **kwargs: Any,
) -> httpx.Response:
    for attempt in range(max_retries + 1):
        try:
            response = client.request(method, url, **kwargs)
        except httpx.RequestError as exc:
            if attempt >= max_retries:
                raise CloudflareError(f"Cloudflare request failed: {exc}") from exc
            sleeper(float(2**attempt))
            continue
        if response.status_code not in RETRYABLE_STATUS or attempt >= max_retries:
            return response
        retry_after = response.headers.get("Retry-After")
        try:
            delay = max(0.0, float(retry_after)) if retry_after else float(2**attempt)
        except ValueError:
            delay = float(2**attempt)
        sleeper(delay)
    raise CloudflareError("Cloudflare request retry loop ended unexpectedly")

File: src/http.rs
use crate::error::{BlockholeError, Result};
use reqwest::blocking::{Client, Response};
use std::thread::sleep;
use std::time::Duration;
pub fn request(
    client: &Client,
    method: reqwest::Method,
    url: &str,
    retries: usize,
    body: Option<serde_json::Value>,
) -> Result<Response> {
    for attempt in 0..=retries {
        let mut request = client.request(method.clone(), url);
        if let Some(ref json) = body {
            request = request.json(json);
        }
        match request.send() {
            Ok(response)
                if !matches!(
                    response.status().as_u16(),
                    408 | 429 | 500 | 502 | 503 | 504
                ) || attempt == retries =>
            {
                return Ok(response);
            }
            Ok(response) => {
                let delay = response
                    .headers()
                    .get("Retry-After")
                    .and_then(|v| v.to_str().ok())
                    .and_then(|v| v.parse::<f64>().ok())
                    .unwrap_or(2_f64.powi(attempt as i32));
                sleep(Duration::from_secs_f64(delay.max(0.0)));
            }
            Err(err) if attempt == retries => {
                return Err(BlockholeError::Cloudflare(format!("request failed: {err}")));
            }
            Err(_) => sleep(Duration::from_secs_f64(2_f64.powi(attempt as i32))),
        }
    }
    Err(BlockholeError::Cloudflare(
        "retry loop ended unexpectedly".into(),
    ))
}
